doc/ir-semantics.text

```diff
@@ -121,49 +121,103 @@ declares i =
+drop all same-named bindings before a declaration.
+
+¹: the tightest possible assignment can be defined impredicatively as
+the pointwise intersection of all valid judgments, which is
+well-defined given that valid scopes are bounded by the set of
+variables defined in the program.
+
+  ∀(v → M) ∈ P,   A, P ⊨d v → M
+  ----------------------------
+            A, P ⊨d
+
+  ∀(l → i) ∈ M,   A, P ⊨d v, l → i
+  -------------------------------
+         A, P ⊨d v → M
+
+  S := A(v,l)
+  requires i ⊆ S
+  ∀l* ∈ pred(P(v),l), A(v,l*) ⊇ S
+  ∀l' ∈ succ(P(v),l), S ∪ declares i ⊇ A(v,l')
+  A, P ⊨d i : S
+  ----------------------------------------
+  A, P ⊨d v, l → i
+
+The judgment (A, P ⊢ i : S) below is the instruction-specific judgment
+that checks the scope restrictions that are specific to each
+instruction, rather than the general control-flow constraints and the
+generic requires/declares handling.
+
+  ----------------
+  A, P ⊨d start : ∅
+
+              A(v, l) = x*
+  ----------------------------------
+  A, P ⊨d osr(e, v, l, (x = ee*)) : S
+
+  ---------------
+  A, P ⊨d stop : ∅
+
+
+## Algorithmic scoping judgment
+
+(A, P ⊨a), (A, P ⊨a v → M), (A, P ⊨a i : S): as in the declarative
+scoping judgment. Only (A, P ⊨a v, l → i) differs.
+
+  S := A(v,l) \ declares i
+  (⋂_{l* ∈ pred(P(v),l)} A(v,l*)) = S ∪ requires i
+  A, P ⊨a i : S
+  ------------------------------------------------
+  A, P ⊨a v, l → i
+
+There is a fundamental difference in the role of the scope assignment
+A between the declarative and algorithmic judgments.
+
+In the declarative judgment, A(v,l) points to the variables
+*available* in the *ingoing* scope of the instruction at location
+(v,l): the scope that is available when arriving to the instruction
+from a predecessor. The variables that are available in the outgoing
+scope can be computed from this value as (A(v,l) ∪ declares i).
+
+In the algorithmic judgment, A(v,l) points to the variables that are
+*required* in the *outgoing* scope of the instruction at location
+(v,l): the scope that is required when leaving the instruction to
+a successor. The variables required in the ingoing scope can be
+computed from this value as (A(v,l) \ declares i).
+
+In the declarative judgment, the relation between an instruction scope
+and its neighbors is under-specified: a predecessor's outgoing scope
+may have more variables than our ingoing scope, and any successor's
+ingoing scope may have less variable than our outgoing scope.
+
+In the algorithmic judgment, the relation between the scopes of an
+instruction and its neighbors is deterministic: an instruction's
+ingoing scope is exactly the variables that are common among all its
+predecessors. Implicitly dropping a variable that will never be used
+again is possible: this is done by choosing a A(v,l) that is smaller
+than the instruction's ingoing set plus (declares i).
+
+Problem: I think there is a problem with the notion of outgoing scope
```

Comment:
Right, both our algorithmic judgment do less than what our implementation did.

Actually it makes sense, because the only way it could be minimal is by requiring both start and end to have an empty scope (and not like now: either of them).

Should we remove the algorithmic version (both) for now? They seem broken.